sycl-fusion/test/materializer/multi_type.ll

```diff
@@ -0,0 +1,82 @@
+; RUN: --mtriple amdgcn-amd-amdhsa -sycl-materializer-debug-value-size=256\
+; RUN: -passes=sycl-spec-const-materializer -S %s | FileCheck\
+; RUN: --check-prefix=CHECK-MATERIALIZER %s %}
+
+; RUN: %if cuda %{ opt -load-pass-plugin %shlibdir/SYCLKernelFusion%shlibext\
+; RUN: --mtriple nvptx64-nvidia-cuda -sycl-materializer-debug-value-size=256\
+; RUN: -passes=sycl-spec-const-materializer -S %s | FileCheck\
+; RUN: --check-prefix=CHECK-MATERIALIZER %s %}
+
+; RUN: %if hip_amd %{ opt -load-pass-plugin %shlibdir/SYCLKernelFusion%shlibext\
+; RUN: --mtriple amdgcn-amd-amdhsa -sycl-materializer-debug-value-size=256\
+; RUN: -passes=sycl-spec-const-materializer,early-cse -S %s | FileCheck\
+; RUN: --check-prefix=CHECK-MATERIALIZER-CSE %s %}
+
+; RUN: %if cuda %{ opt -load-pass-plugin %shlibdir/SYCLKernelFusion%shlibext\
+; RUN: --mtriple nvptx64-nvidia-cuda -sycl-materializer-debug-value-size=256\
+; RUN: -passes=sycl-spec-const-materializer,early-cse -S %s | FileCheck\
+; RUN: --check-prefix=CHECK-MATERIALIZER-CSE %s %}
+
+source_filename = "multi_type.ll"
+target datalayout = "e-i64:64-i128:128-v16:16-v32:32-n16:32:64"
+
+; The same logic as in the basic.ll, but with more complicated types; an array
+; of struct with multiple members. This is important as the pass has to
+; correctly set the type of offsetted memory region corresponding to
+; specialization constant (see getConstantOfType in the pass).
+; For CHECK-MATERIALIZER-CSE expect literal store only.
+
+;CHECK-MATERIALIZER: @SpecConsBlob___test_kernel_0 = weak_odr addrspace(4) constant i32
+;CHECK-MATERIALIZER: @SpecConsBlob___test_kernel_1 = weak_odr addrspace(4) constant %"struct.std::array"
+;CHECK-MATERIALIZER: @SpecConsBlob___test_kernel_2 = weak_odr addrspace(4) constant [2 x i32]
+
+%"struct.std::array" = type { [5 x %struct.ThreePrimitives] }
+%struct.ThreePrimitives = type <{ double, i64, half }>
+
+;CHECK: __test_kernel
+define weak_odr protected amdgpu_kernel void @__test_kernel(ptr addrspace(1) noundef align 4 %out, ptr addrspace(1) noundef align 1 %_arg__specialization_constants_buffer) {
+entry:
+  ;CHECK-MATERIALIZER-CSE-NOT: load
+  ;CHECK-MATERIALIZER-CSE: store double 1.100000e+01, ptr addrspace(1) %out
+
+  ;CHECK-MATERIALIZER-CSE-NOT: addrspacecast ptr addrspace(1) %_arg__specialization_constants_buffer to ptr
+  %0 = addrspacecast ptr addrspace(1) %_arg__specialization_constants_buffer to ptr
+  %gep = getelementptr i8, ptr %0, i32 0
+  %bc = bitcast ptr %gep to ptr
+  ;CHECK-MATERIALIZER: load i32, ptr addrspace(4) @SpecConsBlob___test_kernel_0
+  %load = load i32, ptr %bc, align 4
+
+  %gep2 = getelementptr i8, ptr %0, i32 4
+  %bc2 = bitcast ptr %gep2 to ptr
+  ;CHECK-MATERIALIZER: load [2 x i32], ptr addrspace(4) @SpecConsBlob___test_kernel_2
+  %load2 = load [2 x i32], ptr %bc2, align 4
+  %extract1 = extractvalue [2 x i32] %load2, 0
+  %extract2 = extractvalue [2 x i32] %load2, 1
+
+  ; See the TODO above, this should be const propagated
```

Comment:
Which `TODO` does this refer to?

Reply:
Sorry, that's development detritus, will remove.